logbat/src/main/java/info/logbat/log/domain/Log.java

```diff
@@ -0,0 +1,43 @@
+package info.logbat.log.domain;
+
+import info.logbat.log.domain.enums.Level;
+import info.logbat.log.domain.values.LogData;
+import java.time.LocalDateTime;
+import lombok.Getter;
+
+@Getter
+public class Log {
+
+  private final Long logId;
+  private final Long applicationId;
+  private final Level level;
+  private final LogData logData;
+  private final LocalDateTime timestamp;
+
+  public Log(Long applicationId, String level, String logData, LocalDateTime timestamp) {
+    this(null, applicationId, level, logData, timestamp);
+  }
+
+  public Log(Long logId, Long applicationId, String level, String logData,
+      LocalDateTime timestamp) {
+    this.logId = logId;
+    validateApplicationId(applicationId);
+    this.applicationId = applicationId;
+    this.level = Level.from(level);
+    this.logData = new LogData(logData);
+    validateTimestamp(timestamp);
+    this.timestamp = timestamp;
+  }
```

Comment:
DB 조회 이후 객체 반환과 입력을 위한 처리를 일원화 하기 위한것으로 보입니다! 하지만 DB 조회때도 불필요한 검증 로직이 들어가 있는게 아쉬운거 같아요..

```suggestion
  public Log(Long applicationId, String level, String logData, LocalDateTime timestamp) {
    validateApplicationId(applicationId);
    validateTimestamp(timestamp);
    this(null, applicationId, level, logData, timestamp);
  }

  public Log(Long logId, Long applicationId, String level, String logData,
      LocalDateTime timestamp) {
    this.logId = logId;
    this.applicationId = applicationId;
    this.level = Level.from(level);
    this.logData = new LogData(logData);
    this.timestamp = timestamp;
  }
```

이런 식은 어떨까요?

Reply:
저도 해당 의견에 동의하는데, 위와 같이 구현하니

```
Call to 'this()' must be first statement in constructor body
```
라는 컴파일 오류가 발생하네요!! 어떻게 하는게 좋을까요?

Reply:
생성자에서 다른생성자를 호출할때 항상 제일 처음에 와야하는 규칙때문이네요

코드 중복처럼 보일 수 있곘지만 직접 필드 값에 넣어주는건 어떨까요?
```java
  public Log(Long applicationId, String level, String logData, LocalDateTime timestamp) {
    validateApplicationId(applicationId);
    validateTimestamp(timestamp);
    this.logId = null;
    this.applicationId = applicationId;
    this.level = Level.from(level);
    this.logData = new LogData(logData);
    this.timestamp = timestamp;
  }
```

Reply:
아 JDK 21에서 개선된걸 깜빡했습니다!
그렇다면 검증 메서드가 직접 값을 반환해 주는 건 어떨까요?

Reply:
@tidavid1 님이 말씀하시는 부분이 DB에 저장된 값을 읽을 때의 유효성 검사가 빠져도 되는 점을 언급하신 것 같아요!! 

두 가지 방법이 가능할 것 같은데
- 말씀하신 것 처럼 this()를 호출하지 않고, 필드에 할당하는 코드를 중복 작성해서 처리
- 기존 처럼 DB에 저장된 값을 읽을 때에도 유효성 검사 진행

두 방법 중에 어떻게 하는 것이 좋을까요?

Reply:
이 방법도
```
Cannot reference 'Log.validateApplicationId()' before supertype constructor has been called
```
라고 하네요!! ㅠ

유효성 검사를 반드시 빼는 게 낫다면 필드 할당 로직을 중복하여 작성하면 될 것 같아요!!

Reply:
제가 이해한대로면
static 메서드로 만들어야하는게 맞을까요?

Reply:
```java
/**
   * 새로운 로그를 생성할 때 사용하는 생성자
   *
   * @param applicationId 어플리케이션 ID
   * @param level         로그 레벨
   * @param logData       로그 데이터
   * @param timestamp     로그 타임스탬프
   */
  public Log(Long applicationId, String level, String logData, LocalDateTime timestamp) {
    validateApplicationId(applicationId);
    validateTimestamp(timestamp);
    this.logId = null;
    this.applicationId = applicationId;
    this.level = Level.from(level);
    this.logData = LogData.from(logData);
    this.timestamp = timestamp;
  }

  /**
   * DB에서 로그를 가져올 때 사용하는 생성자
   *
   * @param logId        로그 ID
   * @param applicationId 어플리케이션 ID
   * @param level        로그 레벨
   * @param logData      로그 데이터
   * @param timestamp    로그 타임스탬프
   */
  public Log(Long logId, Long applicationId, String level, String logData,
      LocalDateTime timestamp) {
    this.logId = logId;
    this.applicationId = applicationId;
    this.level = Level.from(level);
    this.logData = LogData.from(logData);
    this.timestamp = timestamp;
  }
```

이런 느낌도 괜찮을 것 같아요!!

Reply:
```java
  public static Log of(Long applicationId, String level, String logData, LocalDateTime timestamp) {
    validateApplicationId(applicationId);
    validateTimestamp(timestamp);
    return new Log(null,  applicationId, level, logData, timestamp);
  }

  public Log(Long logId, Long applicationId, String level, String logData,
      LocalDateTime timestamp) {
    this.logId = logId;
    validateApplicationId(applicationId);
    this.applicationId = applicationId;
    this.level = Level.from(level);
    this.logData = new LogData(logData);
    validateTimestamp(timestamp);
    this.timestamp = timestamp;
  }

  private static void validateApplicationId(Long applicationId) {
    if (applicationId == null) {
      throw new IllegalArgumentException("applicationId는 null이 될 수 없습니다.");
    }
  }

  private static void validateTimestamp(LocalDateTime timestamp) {
    if (timestamp == null) {
      throw new IllegalArgumentException("timestamp는 null이 될 수 없습니다.");
    }
  }
```

이런 느낌은 어떨까요?
메서드명은 `of`, `from`, `create`선택해서요